lightning/src/ln/peer_handler.rs

```diff
@@ -1646,6 +1687,25 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 
 			for (descriptor, peer_mutex) in peers.iter() {
 				self.do_attempt_write_data(&mut (*descriptor).clone(), &mut *peer_mutex.lock().unwrap());
+
+				// Only see if we have room for onion messages after we've written all channel messages, to
+				// ensure the latter take priority.
+				loop {
+					let (peer_node_id, om_buffer_slots_avail) = {
+						let peer = peer_mutex.lock().unwrap();
+						if let Some(peer_node_id) = peer.their_node_id {
+							(peer_node_id, peer.onion_message_buffer_slots_available())
+						} else { break; }
+					};
+					if om_buffer_slots_avail == 0 { break; }
+					let onion_msgs = self.message_handler.onion_message_handler.next_onion_messages_for_peer(
+						peer_node_id, om_buffer_slots_avail);
+					if onion_msgs.len() == 0 { break; }
+					for msg in onion_msgs {
+						self.enqueue_message(&mut *get_peer_for_forwarding!(&peer_node_id), &msg);
+					}
+					self.do_attempt_write_data(&mut (*descriptor).clone(), &mut *peer_mutex.lock().unwrap());
+				}
```

Comment:
can be possible extracted to write_onion_messages() / send_onion_msgs(), otherwise this logic is partially eclipsing code line for sending channel_msgs

Reply:
It changed quite a bit, let me know if this still applies 😬